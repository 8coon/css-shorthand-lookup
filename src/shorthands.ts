const shorthands = {
	'animation': [
		'-name',
		'-duration',
		'-timing-function',
		'-delay',
		'-iteration-count',
		'-direction',
		'-fill-mode',
		'-play-state',
	],
	'background': [
		'-clip',
		'-color',
		'-image',
		'-origin',
		'-position',
		'-repeat',
		'-size',
		'-attachment',
	],
	'columns': [
		'-count',
		'-width',
	],
	'flex': [
		'-grow',
		'-shrink',
		'-basis',
	],
	'flex-flow': [
		'flex-direction',
		'flex-wrap',
	],
	'font': [
		'-style',
		'-variant',
		'-weight',
		'-stretch',
		'-size',
		'line-height',
		'-family',
	],
	'grid': [
		'-template-rows',
		'-template-columns',
		'-template-areas',
		'-auto-rows',
		'-auto-columns',
		'-auto-flow',
		'-column-gap',
		'-row-gap',
		'column-gap',
		'row-gap',
	],
	'grid-area': [
		'grid-row-start',
		'grid-column-start',
		'grid-row-end',
		'grid-column-end',
	],
	'grid-column': [
		'-start',
		'-end',
	],
	'grid-row': [
		'-start',
		'-end',
	],
	'grid-template': [
		'-areas',
		'-columns',
		'-rows',
	],
	'list-style': [
		'-image',
		'-position',
		'-type',
	],
	'offset': [
		'-anchor',
		'-distance',
		'-path',
		'-position',
		'-rotate',
	],
	'place-content': [
		'align-content',
		'justify-content',
	],
	'place-items': [
		'align-items',
		'justify-items',
	],
	'place-self': [
		'align-self',
		'justify-self',
	],
	'text-decoration': [
		'-line',
		'-color',
		'-style',
		'-thickness',
	],
	'transition': [
		'-property',
		'-duration',
		'-timing-function',
		'-delay',
	],

	// Generated at run-time
	'border': [],
	'border-radius': [],
	'column-rule': [],
	'margin': [],
	'outline': [],
	'padding': [],
}

const topSide = 'top';
const bottomSide = 'bottom';
const rightSide = 'right';
const leftSide = 'left';

const sides = [topSide, rightSide, bottomSide, leftSide];
const verticalSides = [topSide, bottomSide];
const horizontalSides = [rightSide, leftSide];

const borderCommonProps = ['width', 'style', 'color'];

for (const prop of borderCommonProps) {
	const propAffix = `-${prop}`;

	shorthands[`border${propAffix}`] = [];
	shorthands['border'].push(propAffix);
	shorthands['column-rule'].push(propAffix);
	shorthands['outline'].push(propAffix);
}

for (const side of sides) {
	const sideAffix = `-${side}`;

	shorthands[`border${sideAffix}`] = shorthands['border'];
	shorthands['margin'].push(sideAffix);
	shorthands['padding'].push(sideAffix);

	for (const prop of borderCommonProps) {
		shorthands[`border-${prop}`].push(`border${sideAffix}-${prop}`);
	}
}

for (const verticalSide of verticalSides) {
	for (const horizontalSide of horizontalSides) {
		shorthands['border-radius'].push(`border-${verticalSide}-${horizontalSide}-radius`);
	}
}

// Expanding all props and building a reverse lookup map
const shorthandKeys = Object.keys(shorthands);
export const shorthandReverseMap: Record<string, string[]> = {};

for (const shorthand of shorthandKeys) {
	for (const partialProp of shorthands[shorthand]) {
		const fullProp = partialProp.startsWith('-')
			? shorthand + partialProp
			: partialProp;

		shorthandReverseMap[fullProp] = shorthandReverseMap[fullProp] ?? [];
		shorthandReverseMap[fullProp].push(shorthand);
	}
}
