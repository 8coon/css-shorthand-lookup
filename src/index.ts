import {shorthandReverseMap} from './shorthands';

export function lookupShorthands(fullPropName: string, rule?: CSSStyleRule): string[] {
	const lookupResult = shorthandReverseMap[fullPropName] ?? [];

	if (rule) {
		return lookupResult
			.filter((prop) => rule.style.getPropertyValue(prop))
			.sort((a, b) => rule.cssText.indexOf(a) - rule.cssText.indexOf(b));
	}

	return lookupResult;
}
