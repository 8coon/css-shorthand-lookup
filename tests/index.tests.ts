import { shorthandReverseMap } from "../src/shorthands";
import { lookupShorthands } from "../src";
import { createStyleRule } from "./createStyleRule";

describe("lookupShorthands", () => {
	test("lookup map matches snapshot", () => {
		expect(shorthandReverseMap).toMatchSnapshot();
	});

	test("lookup function returns base property", () => {
		expect(lookupShorthands("border-top-color")).toEqual([
			"border-color",
			"border-top",
		]);
	});

	test("lookup function returns nothing is no base property exists", () => {
		expect(lookupShorthands("border")).toEqual([]);
	});

	test("lookup function returns base properties that exist in the CSSRule", () => {
		const rule = createStyleRule(`
			.class {
				border: 1px solid var(--border-color);
			}
		`);

		expect(lookupShorthands("border-color", rule)).toEqual(["border"]);
	});

	test("lookup function sorts props based on their order in css", () => {
		const rule = createStyleRule(`
			.class {
				grid-column: inherit;
				grid-area: inherit;
			}
		`);

		expect(lookupShorthands("grid-column-start", rule)).toEqual([
			"grid-column",
			"grid-area",
		]);
	});

	test("lookup function returns nothing if base property does not exist on a rule", () => {
		const rule = createStyleRule(`
			.class {
				border: 1px solid var(--border-color);
			}
		`);

		expect(lookupShorthands("grid-column-start", rule)).toEqual([]);
	});
});
