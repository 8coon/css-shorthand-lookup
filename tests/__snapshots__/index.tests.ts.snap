// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lookupShorthands lookup map matches snapshot 1`] = `
Object {
  "align-content": Array [
    "place-content",
  ],
  "align-items": Array [
    "place-items",
  ],
  "align-self": Array [
    "place-self",
  ],
  "animation-delay": Array [
    "animation",
  ],
  "animation-direction": Array [
    "animation",
  ],
  "animation-duration": Array [
    "animation",
  ],
  "animation-fill-mode": Array [
    "animation",
  ],
  "animation-iteration-count": Array [
    "animation",
  ],
  "animation-name": Array [
    "animation",
  ],
  "animation-play-state": Array [
    "animation",
  ],
  "animation-timing-function": Array [
    "animation",
  ],
  "background-attachment": Array [
    "background",
  ],
  "background-clip": Array [
    "background",
  ],
  "background-color": Array [
    "background",
  ],
  "background-image": Array [
    "background",
  ],
  "background-origin": Array [
    "background",
  ],
  "background-position": Array [
    "background",
  ],
  "background-repeat": Array [
    "background",
  ],
  "background-size": Array [
    "background",
  ],
  "border-bottom-color": Array [
    "border-color",
    "border-bottom",
  ],
  "border-bottom-left-radius": Array [
    "border-radius",
  ],
  "border-bottom-right-radius": Array [
    "border-radius",
  ],
  "border-bottom-style": Array [
    "border-style",
    "border-bottom",
  ],
  "border-bottom-width": Array [
    "border-width",
    "border-bottom",
  ],
  "border-color": Array [
    "border",
  ],
  "border-left-color": Array [
    "border-color",
    "border-left",
  ],
  "border-left-style": Array [
    "border-style",
    "border-left",
  ],
  "border-left-width": Array [
    "border-width",
    "border-left",
  ],
  "border-right-color": Array [
    "border-color",
    "border-right",
  ],
  "border-right-style": Array [
    "border-style",
    "border-right",
  ],
  "border-right-width": Array [
    "border-width",
    "border-right",
  ],
  "border-style": Array [
    "border",
  ],
  "border-top-color": Array [
    "border-color",
    "border-top",
  ],
  "border-top-left-radius": Array [
    "border-radius",
  ],
  "border-top-right-radius": Array [
    "border-radius",
  ],
  "border-top-style": Array [
    "border-style",
    "border-top",
  ],
  "border-top-width": Array [
    "border-width",
    "border-top",
  ],
  "border-width": Array [
    "border",
  ],
  "column-gap": Array [
    "grid",
  ],
  "column-rule-color": Array [
    "column-rule",
  ],
  "column-rule-style": Array [
    "column-rule",
  ],
  "column-rule-width": Array [
    "column-rule",
  ],
  "columns-count": Array [
    "columns",
  ],
  "columns-width": Array [
    "columns",
  ],
  "flex-basis": Array [
    "flex",
  ],
  "flex-direction": Array [
    "flex-flow",
  ],
  "flex-grow": Array [
    "flex",
  ],
  "flex-shrink": Array [
    "flex",
  ],
  "flex-wrap": Array [
    "flex-flow",
  ],
  "font-family": Array [
    "font",
  ],
  "font-size": Array [
    "font",
  ],
  "font-stretch": Array [
    "font",
  ],
  "font-style": Array [
    "font",
  ],
  "font-variant": Array [
    "font",
  ],
  "font-weight": Array [
    "font",
  ],
  "grid-auto-columns": Array [
    "grid",
  ],
  "grid-auto-flow": Array [
    "grid",
  ],
  "grid-auto-rows": Array [
    "grid",
  ],
  "grid-column-end": Array [
    "grid-area",
    "grid-column",
  ],
  "grid-column-gap": Array [
    "grid",
  ],
  "grid-column-start": Array [
    "grid-area",
    "grid-column",
  ],
  "grid-row-end": Array [
    "grid-area",
    "grid-row",
  ],
  "grid-row-gap": Array [
    "grid",
  ],
  "grid-row-start": Array [
    "grid-area",
    "grid-row",
  ],
  "grid-template-areas": Array [
    "grid",
    "grid-template",
  ],
  "grid-template-columns": Array [
    "grid",
    "grid-template",
  ],
  "grid-template-rows": Array [
    "grid",
    "grid-template",
  ],
  "justify-content": Array [
    "place-content",
  ],
  "justify-items": Array [
    "place-items",
  ],
  "justify-self": Array [
    "place-self",
  ],
  "line-height": Array [
    "font",
  ],
  "list-style-image": Array [
    "list-style",
  ],
  "list-style-position": Array [
    "list-style",
  ],
  "list-style-type": Array [
    "list-style",
  ],
  "margin-bottom": Array [
    "margin",
  ],
  "margin-left": Array [
    "margin",
  ],
  "margin-right": Array [
    "margin",
  ],
  "margin-top": Array [
    "margin",
  ],
  "offset-anchor": Array [
    "offset",
  ],
  "offset-distance": Array [
    "offset",
  ],
  "offset-path": Array [
    "offset",
  ],
  "offset-position": Array [
    "offset",
  ],
  "offset-rotate": Array [
    "offset",
  ],
  "outline-color": Array [
    "outline",
  ],
  "outline-style": Array [
    "outline",
  ],
  "outline-width": Array [
    "outline",
  ],
  "padding-bottom": Array [
    "padding",
  ],
  "padding-left": Array [
    "padding",
  ],
  "padding-right": Array [
    "padding",
  ],
  "padding-top": Array [
    "padding",
  ],
  "row-gap": Array [
    "grid",
  ],
  "text-decoration-color": Array [
    "text-decoration",
  ],
  "text-decoration-line": Array [
    "text-decoration",
  ],
  "text-decoration-style": Array [
    "text-decoration",
  ],
  "text-decoration-thickness": Array [
    "text-decoration",
  ],
  "transition-delay": Array [
    "transition",
  ],
  "transition-duration": Array [
    "transition",
  ],
  "transition-property": Array [
    "transition",
  ],
  "transition-timing-function": Array [
    "transition",
  ],
}
`;
